File: py/auth_callback.py
#!/usr/bin/env python3

import cgi
import urllib.parse
import urllib.request
import base64
import json
from http import cookies

form = cgi.FieldStorage()

code = form["code"].value

url = "https://www.reddit.com/api/v1/access_token"
values = {
    "grant_type": "authorization_code",
    "code": code,
    "redirect_uri": "http://localhost:8000/py/auth_callback.py"
}

password_str = base64.b64encode(b"ntFDwUGcEC-X4A:XWCtr0HSC-E2tKMrK5W5qgaGZqQ")

headers = {
    "Content-Type" : "application/x-www-form-urlencoded",
    "Authorization": "Basic " + str(password_str.decode("ascii"))
}

data = urllib.parse.urlencode(values)
data = data.encode("utf-8")

req = urllib.request.Request(url, data, headers)

response = urllib.request.urlopen(req)

response_json = json.loads(response.read().decode("utf-8"))
token = response_json["access_token"]

cookie = cookies.SimpleCookie()
cookie["oat"] = token
cookie["oat"]["expires"] = int(response_json["expires_in"])
cookie["oat"]["path"] = "/"

print(str(cookie))
print("Content-Type: text/html\n")
print("<meta http-equiv='Refresh' content='0; url=/wall/' />")

File: routes/auth/authCallback.js
var express = require("express");
var router = express.Router();

router.get("/", function(req, res, next) {
    var request = require("request");
    var base64 = require("base-64");
    var utf8 = require("utf8");

    var URL = "https://www.reddit.com/api/v1/access_token";

    var code = req.query.code;
    var error = req.query.error;
    if (error) {
        console.log(error);
        res.sendStatus(500);
    }

    var secret = process.env.REDDIT_API_SECRET;
    if (!secret) {
        var err = new Error("No Reddit API app secret defined");
        err.status = 500;
        next(err);
    }
    var passwordB64 = base64.encode(utf8.encode(["jQ9e-THHchxOZg", secret].join(":")));

    var form = {
        grant_type: "authorization_code",
        code: code,
        redirect_uri: "http://localhost:3000/auth/auth_callback"
    };
    var headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": "Basic " + passwordB64,
        "User-Agent": "web alienwall v0.0.1 (by /u/thedonkeypie)"
    };

    request.post({
        url: URL,
        headers: headers,
        form: form
    }, function(error, response, body) {
        if (!error && response.statusCode.toString()[0] === "2") {
            var now = new Date();

            var responseBody = JSON.parse(body);
            console.log(responseBody);

            if (!responseBody.error) {
                var accessToken = responseBody.access_token;
                var tokenExpiry = new Date(now.getTime() + Number(responseBody.expires_in) * 1000);
                var refreshToken = responseBody.refresh_token;

                var tenYearsInTheFuture = new Date(now.getTime() + 315569260000);

                res.cookie("access_token", accessToken, { expires: tokenExpiry });
                res.cookie("refresh_token", refreshToken, { expires: tenYearsInTheFuture });
                res.cookie("authdate", now.getTime(), { expires: tenYearsInTheFuture });

                res.append("Content-Type", "text/html");
                res.redirect("/wall/");
            } else {
                res.sendStatus(500);
            }
        } else {
            res.sendStatus(500);
        }
    });
});

module.exports = router;
